pkg/virt-handler/notify-server/server.go

```diff
@@ -154,7 +155,23 @@ func RunServer(virtShareDir string, stopChan chan struct{}, c chan watch.Event,
 	case <-done:
 		log.Log.Info("notify server done")
 	case <-stopChan:
-		grpcServer.Stop()
+		gracefulShutdownTimeout := 10
+
+		isStopped := make(chan struct{})
+		go func() {
+			grpcServer.GracefulStop()
+			close(isStopped)
+		}()
```

Comment:
What assures that this go routine finishes and does not leak by blocking on `GracefulStop`?

Reply:
Please see the conversion we had in this PR (its in couple comments),
we added the timer since GracefulStop is unbounded, in case we have a bug with our Handler
(split the flow until grpc serveStream (1) and forward, where our Handler takes place (2))
we dont want to wait forever, and find it out in weird ways as David suggested,
this is why there is the timer and then we fallback to Stop.
No matter what mechanism we would implement that would wait for connections to finish (2), would have same, theoretical only for now, possible hanging, its a precaution, not based on true problem we saw.

Moreover, we discussed that ERROR of clients shouldn't stop the server upon application layer ERROR, so this mechanism would not be trigger as it was triggered every now and then.

Reply:
Thank you for the info but this is not what I asked.
I asked what assures that this go-routine will terminate. Or in other words, what assures that the `GracefulStop` returns.

Reply:
once it collected all the connections it will return, docs says that , code tells that.
in case it doesnt , this is why we added the timer

Reply:
If the timer kicks in, it will call a forceful `Stop`, what will happen then to this call? Can you depend on `Stop` causing `GracefulStop` to return?

Reply:
yes, it does, there is a code in Stop that release it

Reply:
IMO depending on the implementation details is dangerous. I would expect to see this explicitly mentioned in the API documentation, otherwise they may just change the code and not even think about such a scenario.
As this is also non trivial, I would even add a comment in the commit message about this so future us will know we examine it already.

Lets see what @AlonaKaplan and others think about this.

Reply:
its written in the code documents, and i verified on the code and with the community

Reply:
> its written in the code documents, and i verified on the code and with the community.
> this is the meaning of GracefulStop

The only documentation I see is [this](https://godoc.org/google.golang.org/grpc#Server.GracefulStop).
Please ref the documentation you think is relevant.

> Menghan Li said
> "GracefulStop() doesn't add limitation to how streams can be used. Your spawned goroutines can use stream as normal. The server waits for the service handler to return."

How is this related to `Stop()` causing this one to return?

> Power of a well maintained library should be used. our code is built find for that.

I do not understand how the library stability has anything to do with my question.
You are using an assumption and I ask you to confirm it is correct and that the library maintainers declared it as an explicit behavior.

Reply:
i dont assume, i just talked generally on GracefulStop instead how stop stops GracefulStop, 
i will enclose code snippet from the code, but not at this very moment, need to handle something before please

Reply:
The `Stop` cleans all the connection (even the running ones) and [signals](https://github.com/kubevirt/kubevirt/blob/8d8baaf19e4f10d75d91072daf2b566937d00b0c/vendor/google.golang.org/grpc/server.go#L1409) about it. So the `GracefulStop` will [get the signal](https://github.com/kubevirt/kubevirt/blob/8d8baaf19e4f10d75d91072daf2b566937d00b0c/vendor/google.golang.org/grpc/server.go#L1462) and see there are no connection so it will move on and stop its `wait` for the connections to be clean.

Reply:
Thank you @AlonaKaplan !